File: java/src/main/java/network/qorechain/pqc/Pqc.java
package network.qorechain.pqc;

import java.nio.ByteBuffer;
import java.security.SecureRandom;

import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.SecretWithEncapsulation;
import org.bouncycastle.crypto.digests.SHAKEDigest;
import org.bouncycastle.crypto.params.ParametersWithRandom;
import org.bouncycastle.pqc.crypto.mldsa.MLDSAKeyGenerationParameters;
import org.bouncycastle.pqc.crypto.mldsa.MLDSAKeyPairGenerator;
import org.bouncycastle.pqc.crypto.mldsa.MLDSAParameters;
import org.bouncycastle.pqc.crypto.mldsa.MLDSAPrivateKeyParameters;
import org.bouncycastle.pqc.crypto.mldsa.MLDSAPublicKeyParameters;
import org.bouncycastle.pqc.crypto.mldsa.MLDSASigner;
import org.bouncycastle.pqc.crypto.mlkem.MLKEMExtractor;
import org.bouncycastle.pqc.crypto.mlkem.MLKEMGenerator;
import org.bouncycastle.pqc.crypto.mlkem.MLKEMKeyGenerationParameters;
import org.bouncycastle.pqc.crypto.mlkem.MLKEMKeyPairGenerator;
import org.bouncycastle.pqc.crypto.mlkem.MLKEMParameters;
import org.bouncycastle.pqc.crypto.mlkem.MLKEMPrivateKeyParameters;
import org.bouncycastle.pqc.crypto.mlkem.MLKEMPublicKeyParameters;

/**
 * QoreChain post-quantum cryptography (Bouncy Castle binding).
 *
 * <p>Standardized NIST primitives:
 * <ul>
 *   <li>ML-DSA (FIPS-204) signatures — levels 44 / 65 / 87 (default 87)</li>
 *   <li>ML-KEM (FIPS-203) key encapsulation — levels 512 / 768 / 1024 (default 1024)</li>
 *   <li>SHAKE-256 (FIPS-202) extendable-output hash</li>
 * </ul>
 *
 * The API mirrors the other language bindings in this repo and is validated
 * against the shared vectors in {@code /vectors}.
 */
public final class Pqc {
    public static final String VERSION = "0.1.0";

    private Pqc() {}

    // ── ML-DSA (FIPS-204) ────────────────────────────────────────────────────

    private static MLDSAParameters mldsaParams(String level) {
        switch (level) {
            case "ml-dsa-44": return MLDSAParameters.ml_dsa_44;
            case "ml-dsa-65": return MLDSAParameters.ml_dsa_65;
            case "ml-dsa-87": return MLDSAParameters.ml_dsa_87;
            default: throw new IllegalArgumentException("unknown ML-DSA level: " + level);
        }
    }

    /**
     * Generate an ML-DSA keypair. Returns {publicKey, secretKey} where secretKey
     * is the 32-byte FIPS-204 key-generation seed (ξ).
     *
     * <p>Bouncy Castle 1.79 represents an ML-DSA private key by its seed; the
     * expanded {@code getEncoded()} form does not round-trip through the
     * {@code (params, byte[])} constructor, so the seed is the portable secret
     * key here. Verification and the cross-language vectors are unaffected
     * (those exchange public keys and signatures, never private keys).
     */
    public static byte[][] mldsaKeygen(String level) {
        MLDSAKeyPairGenerator g = new MLDSAKeyPairGenerator();
        g.init(new MLDSAKeyGenerationParameters(new SecureRandom(), mldsaParams(level)));
        AsymmetricCipherKeyPair kp = g.generateKeyPair();
        byte[] pk = ((MLDSAPublicKeyParameters) kp.getPublic()).getEncoded();
        byte[] sk = ((MLDSAPrivateKeyParameters) kp.getPrivate()).getSeed();
        return new byte[][] {pk, sk};
    }

    /** Sign {@code message} with {@code secretKey} (the 32-byte seed; empty context, FIPS-204 pure). */
    public static byte[] mldsaSign(String level, byte[] secretKey, byte[] message) {
        MLDSAPrivateKeyParameters sk = new MLDSAPrivateKeyParameters(mldsaParams(level), secretKey);
        MLDSASigner signer = new MLDSASigner();
        signer.init(true, new ParametersWithRandom(sk, new SecureRandom()));
        signer.update(message, 0, message.length);
        try {
            return signer.generateSignature();
        } catch (org.bouncycastle.crypto.CryptoException e) {
            throw new IllegalStateException("ML-DSA signing failed", e);
        }
    }

    /** Verify {@code signature} over {@code message} under {@code publicKey}. */
    public static boolean mldsaVerify(String level, byte[] publicKey, byte[] message, byte[] signature) {
        MLDSAPublicKeyParameters pk = new MLDSAPublicKeyParameters(mldsaParams(level), publicKey);
        MLDSASigner signer = new MLDSASigner();
        signer.init(false, pk);
        signer.update(message, 0, message.length);
        return signer.verifySignature(signature);
    }

    // ── ML-KEM (FIPS-203) ────────────────────────────────────────────────────

    private static MLKEMParameters mlkemParams(String level) {
        switch (level) {
            case "ml-kem-512": return MLKEMParameters.ml_kem_512;
            case "ml-kem-768": return MLKEMParameters.ml_kem_768;
            case "ml-kem-1024": return MLKEMParameters.ml_kem_1024;
            default: throw new IllegalArgumentException("unknown ML-KEM level: " + level);
        }
    }

    /** Generate an ML-KEM keypair. Returns {publicKey, secretKey}. */
    public static byte[][] mlkemKeygen(String level) {
        MLKEMKeyPairGenerator g = new MLKEMKeyPairGenerator();
        g.init(new MLKEMKeyGenerationParameters(new SecureRandom(), mlkemParams(level)));
        AsymmetricCipherKeyPair kp = g.generateKeyPair();
        byte[] pk = ((MLKEMPublicKeyParameters) kp.getPublic()).getEncoded();
        byte[] sk = ((MLKEMPrivateKeyParameters) kp.getPrivate()).getEncoded();
        return new byte[][] {pk, sk};
    }

    /** Encapsulate to {@code publicKey}. Returns {ciphertext, sharedSecret}. */
    public static byte[][] mlkemEncapsulate(String level, byte[] publicKey) {
        MLKEMPublicKeyParameters pk = new MLKEMPublicKeyParameters(mlkemParams(level), publicKey);
        SecretWithEncapsulation enc = new MLKEMGenerator(new SecureRandom()).generateEncapsulated(pk);
        return new byte[][] {enc.getEncapsulation(), enc.getSecret()};
    }

    /** Decapsulate {@code ciphertext} with {@code secretKey}. Returns the shared secret. */
    public static byte[] mlkemDecapsulate(String level, byte[] secretKey, byte[] ciphertext) {
        MLKEMPrivateKeyParameters sk = new MLKEMPrivateKeyParameters(mlkemParams(level), secretKey);
        return new MLKEMExtractor(sk).extractSecret(ciphertext);
    }

    // ── SHAKE-256 (FIPS-202) + blockchain helpers ────────────────────────────

    /** SHAKE-256 digest of {@code data} with {@code outLen} bytes. */
    public static byte[] shake256(byte[] data, int outLen) {
        SHAKEDigest d = new SHAKEDigest(256);
        d.update(data, 0, data.length);
        byte[] out = new byte[outLen];
        d.doFinal(out, 0, outLen);
        return out;
    }

    /** SHAKE-256(publicKey) truncated to {@code len} bytes (pay-to-pubkey-hash). */
    public static byte[] pubkeyHash(byte[] publicKey, int len) {
        return shake256(publicKey, len);
    }

    /** Canonical PQC sign-bytes: BE32(len(b0))||b0||BE32(len(auth))||auth. */
    public static byte[] hybridSignBytes(byte[] bodyWithoutPqcExt, byte[] authInfoBytes) {
        ByteBuffer buf = ByteBuffer.allocate(8 + bodyWithoutPqcExt.length + authInfoBytes.length);
        buf.putInt(bodyWithoutPqcExt.length).put(bodyWithoutPqcExt);
        buf.putInt(authInfoBytes.length).put(authInfoBytes);
        return buf.array();
    }
}

File: go/pqc.go
// Package pqc provides QoreChain post-quantum cryptography.
//
// Standardized NIST primitives, wrapping Cloudflare CIRCL + x/crypto:
//   - ML-DSA   (FIPS-204) signatures — levels 44 / 65 / 87 (default 87)
//   - ML-KEM   (FIPS-203) key encapsulation — levels 512 / 768 / 1024 (default 1024)
//   - SHAKE-256 (FIPS-202) extendable-output hash
//
// The API mirrors the other language bindings in this repo and is validated
// against the shared vectors in /vectors.
package pqc

import (
	"encoding/binary"
	"errors"

	mldsa44 "github.com/cloudflare/circl/sign/mldsa/mldsa44"
	mldsa65 "github.com/cloudflare/circl/sign/mldsa/mldsa65"
	mldsa87 "github.com/cloudflare/circl/sign/mldsa/mldsa87"

	mlkem1024 "github.com/cloudflare/circl/kem/mlkem/mlkem1024"
	mlkem512 "github.com/cloudflare/circl/kem/mlkem/mlkem512"
	mlkem768 "github.com/cloudflare/circl/kem/mlkem/mlkem768"

	"github.com/cloudflare/circl/kem"
	"github.com/cloudflare/circl/sign"
	"golang.org/x/crypto/sha3"
)

const Version = "0.1.1"

// SigScheme is a unified ML-DSA signature scheme handle.
type SigScheme struct {
	Name string
	s    sign.Scheme
}

// KemScheme is a unified ML-KEM key-encapsulation scheme handle.
type KemScheme struct {
	Name string
	s    kem.Scheme
}

// ML-DSA (FIPS-204) levels.
var (
	MLDSA44 = SigScheme{"ml-dsa-44", mldsa44.Scheme()}
	MLDSA65 = SigScheme{"ml-dsa-65", mldsa65.Scheme()}
	MLDSA87 = SigScheme{"ml-dsa-87", mldsa87.Scheme()}
	// MLDSA is the QoreChain default (security level 5).
	MLDSA = MLDSA87
)

// ML-KEM (FIPS-203) levels.
var (
	MLKEM512  = KemScheme{"ml-kem-512", mlkem512.Scheme()}
	MLKEM768  = KemScheme{"ml-kem-768", mlkem768.Scheme()}
	MLKEM1024 = KemScheme{"ml-kem-1024", mlkem1024.Scheme()}
	// MLKEM is the QoreChain default (security level 5).
	MLKEM = MLKEM1024
)

// Keygen returns (publicKey, secretKey) for the scheme.
func (sc SigScheme) Keygen() (pub, sec []byte, err error) {
	pk, sk, err := sc.s.GenerateKey()
	if err != nil {
		return nil, nil, err
	}
	pub, _ = pk.MarshalBinary()
	sec, _ = sk.MarshalBinary()
	return pub, sec, nil
}

// KeygenFromSeed deterministically derives (publicKey, secretKey) from the
// 32-byte FIPS-204 xi seed — reproduces the shared /vectors and the other
// language bindings byte-for-byte.
func (sc SigScheme) KeygenFromSeed(seed []byte) (pub, sec []byte, err error) {
	if len(seed) != sc.s.SeedSize() {
		return nil, nil, errors.New("pqc: bad seed length")
	}
	pk, sk := sc.s.DeriveKey(seed)
	pub, _ = pk.MarshalBinary()
	sec, _ = sk.MarshalBinary()
	return pub, sec, nil
}

// Sign signs message with secretKey (empty context, FIPS-204 pure).
// DETERMINISTIC (FIPS-204 section 3.4): same (secretKey, message) always yields
// the same signature — required by QoreChain's on-chain PQC verifier, and how
// the shared /vectors are generated.
func (sc SigScheme) Sign(secretKey, message []byte) ([]byte, error) {
	sk, err := sc.s.UnmarshalBinaryPrivateKey(secretKey)
	if err != nil {
		return nil, err
	}
	return sc.s.Sign(sk, message, nil), nil
}

// Verify checks signature over message under publicKey.
func (sc SigScheme) Verify(publicKey, message, signature []byte) bool {
	pk, err := sc.s.UnmarshalBinaryPublicKey(publicKey)
	if err != nil {
		return false
	}
	return sc.s.Verify(pk, message, signature, nil)
}

// Keygen returns (publicKey, secretKey).
func (kc KemScheme) Keygen() (pub, sec []byte, err error) {
	pk, sk, err := kc.s.GenerateKeyPair()
	if err != nil {
		return nil, nil, err
	}
	pub, _ = pk.MarshalBinary()
	sec, _ = sk.MarshalBinary()
	return pub, sec, nil
}

// Encapsulate to publicKey returns (ciphertext, sharedSecret).
func (kc KemScheme) Encapsulate(publicKey []byte) (ct, ss []byte, err error) {
	pk, err := kc.s.UnmarshalBinaryPublicKey(publicKey)
	if err != nil {
		return nil, nil, err
	}
	return kc.s.Encapsulate(pk)
}

// Decapsulate ciphertext with secretKey returns the shared secret.
func (kc KemScheme) Decapsulate(secretKey, ciphertext []byte) ([]byte, error) {
	sk, err := kc.s.UnmarshalBinaryPrivateKey(secretKey)
	if err != nil {
		return nil, err
	}
	return kc.s.Decapsulate(sk, ciphertext)
}

// Shake256 returns a SHAKE-256 (FIPS-202) digest of data with outLen bytes.
func Shake256(data []byte, outLen int) []byte {
	out := make([]byte, outLen)
	h := sha3.NewShake256()
	h.Write(data)
	h.Read(out)
	return out
}

// PubkeyHash returns SHAKE-256(publicKey) truncated to len bytes (default 20).
func PubkeyHash(publicKey []byte, length int) []byte { return Shake256(publicKey, length) }

// HybridSignBytes is the canonical message a PQC signature covers in QoreChain's
// hybrid-extension scheme: BE32(len(b0))||b0||BE32(len(auth))||auth, where b0 is
// the tx body WITHOUT the PQC extension.
func HybridSignBytes(bodyWithoutPqcExt, authInfoBytes []byte) []byte {
	out := make([]byte, 0, 8+len(bodyWithoutPqcExt)+len(authInfoBytes))
	var be [4]byte
	binary.BigEndian.PutUint32(be[:], uint32(len(bodyWithoutPqcExt)))
	out = append(out, be[:]...)
	out = append(out, bodyWithoutPqcExt...)
	binary.BigEndian.PutUint32(be[:], uint32(len(authInfoBytes)))
	out = append(out, be[:]...)
	out = append(out, authInfoBytes...)
	return out
}
